### ISSUE TITLE
Cancelling `Reload site` prompt still refreshes the page

### ISSUE THREAD

**Issue by bilogic** [jitsi/jitsi-meet#15198]
### What happened?

https://github.com/user-attachments/assets/85f739eb-bb15-4d28-9fb4-ce40eb12bc97

1. Start a recording
2. Press F5, `Reload site` pops up
3. Click `Cancel`
4. Page refreshes, and recording is downloaded
5. I expected the page would not refresh, and recording continues

My jitsi runs in docker using this commit
https://github.com/jitsi/jitsi-meet/commit/24ae69348bdafe7dba229dcc8fa8d943c16e9469 

### Platform

- [X] Chrome (or Chromium based)
- [ ] Firefox
- [ ] Safari
- [ ] Other desktop browser
- [ ] Android browser
- [ ] iOS browser
- [ ] Electron app
- [ ] Android mobile app
- [ ] iOS mobile app
- [ ] Custom app using a mobile SDK

### Browser / app / sdk version

Chrome 129.0.6668.100 (Official Build) (64-bit) / 

### Reproducibility

- [X] The problem is reproducible on meet.jit.si

**Comment by damencho** [comment]
> * Press F5, `Reload site` pops up

We do not have any such reload, this is handled by the browser not jitsi-meet, isn't it?

**Comment by bilogic** [comment]
Yes, the prompt showed probably because jitsi-meet told Chrome that there is some dirty info on the page.

So the issue here is probably because jitsi-meet did not handle the return.

**Comment by damencho** [comment]
Can you show a screenshot of that dialog? There is no such thing handled in jitsi-meet. 
So I don't see what we can do about it.

**Comment by bilogic** [comment]
![image](https://github.com/user-attachments/assets/534447dc-2037-49ba-94b3-1c592f226844)

1. Are you able to repro 1-4 of my OP?
2. If not, which step were you stuck at on meet.jit.si?
3. The issue is not the prompt, the issue is how jitsi incorrectly handled the prompt
4. When `Cancel` is clicked, jitis has to `preventDefault()` and prevent the page from refreshing + not interrupt the recording

https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault
For the record, JS is not my regular language, but I have sufficient understand about this prompt thing.

**Comment by bilogic** [comment]
> There is no such thing handled in jitsi-meet. So I don't see what we can do about it.

Then it might be a case of adding a handler to deal with it

**Comment by damencho** [comment]
When I do the shortcut for reload the page reloads. There is no dialog. This is what I experience.
That's why I ask for screenshot cause there is no such dialog implemented in jitsi-meet and I think this is something external that has nothing ti do with jitsi-meet.

**Comment by bilogic** [comment]
Did you start recording in step 1?

**Comment by bilogic** [comment]
- repro of the issue

https://github.com/user-attachments/assets/85f739eb-bb15-4d28-9fb4-ce40eb12bc97

**Comment by damencho** [comment]
Thanks for the video. You are right, I was skipping that part, sorry and I was not aware of the dialog. We do have it, but still it is the browser one and so the strings in it are not available in jitsi-meet.
https://github.com/jitsi/jitsi-meet/blob/639114f2e19ad2e619ba81d8c6552b2385169c81/react/features/base/conference/middleware.any.ts#L302
I'm not sure you can detect whether the user has clicked cancel or not 
https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event

**Comment by bilogic** [comment]
> I'm not sure you can detect whether the user has clicked cancel or not
> https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event

Yes you can, that is the sole purpose of that dialog, to prevent users from closing their window resulting in a catastrophic disaster, i.e. loss of data

**Comment by bilogic** [comment]
In fact, it would be great if jitsi always pops this dialog for any refresh as I see no use case for a refresh. I reported an issue here https://github.com/jitsi/jitsi-meet-electron/issues/986

**Comment by aaronkvanmeerten** [comment]
Refresh is actually an important part of the meeting experience.  Anytime a client gets into an inconsistent state (connected to new backends, failure in existing connections) that cannot be resolved internally, a reload is triggered to rejoin the meeting.

**Comment by damencho** [comment]
> Yes you can, that is the sole purpose of that dialog, to prevent users from closing their window resulting in a catastrophic disaster, i.e. loss of data

If you can detect when the user clicks Cancel, please share it with us or create a PR. I'm not familiar with that and a simple search shows that this is not possible with current browser API.

**Comment by bilogic** [comment]
ok will do, give me a while

**Comment by damencho** [comment]
The problem is that when the reload confirmation dialog appears if you click cancel the local recording is stopped and saved.
How in the code in that jsfiddle do you determine whether the user has clicked cancel?

Have you been able to check the code I posted above? It is already calling `preventDefault()`.

https://github.com/jitsi/jitsi-meet/blob/639114f2e19ad2e619ba81d8c6552b2385169c81/react/features/base/conference/middleware.any.ts#L306

**Comment by bilogic** [comment]
> How in the code in that jsfiddle do you determine whether the user has clicked cancel?

https://stackoverflow.com/questions/63405982/how-to-know-when-a-page-redirect-or-refresh-action-is-cancelled-javascript

**Comment by damencho** [comment]
Seems there is no reliable way to do that for now.

**Comment by bilogic** [comment]
Can I suggest blocking F5 and ctrl+R for refreshing and allow refreshes only when a user clicks on the browser's refresh button or click on URL and press enter

**Comment by bilogic** [comment]
@SakkyOP 

1. Reloading with only 1 person does not close the meeting (Is this relevant to the OP?)
2. Stopping a recording to deal with a **possible** unload is logical, but very disruptive, especially if the unload is an accidental one, UI are for humans and ought to be designed as such, mistakes are reversible with minimal/0 consequences

After some consideration, I hope you agree that disable F5/ctrl+R is the simple straight forward way to deal with the OP while still leaving some options to refresh.

**Comment by SakkyOP** [comment]
I think a better approach would be to shift the stop recording functionality to take place when the <code>pagehide</code> / <code>visibilitychange</code> event occurs

Also I noticed that the use of <code>returnValue</> is wrong in <code>beforeunload</code> event handler, the prompt is shown when <code>preventDefault()</code> is called which is correct but in legacy versions the prompt appears when the <code>returnValue</code> property is set to a <code>truthy</code> value ( it is set to <code>null</code> in the current version which is <code>falsy</code> )

### This will help:
- avoid removing functionality of reload shortcuts such as <code>F5</code> and <code>ctrl + R</code>
- only stop the recording when the page is actually reloaded by the user ( hence minimal consequence )
- make the reload prompt more reliable and actually work in  various legacy versions as well

### My references:
- [page life cycle](https://developer.chrome.com/docs/web-platform/page-lifecycle-api/image/page-lifecycle-api-state.svg) This was to get a better understanding where and when to use <code>visibilitychange</code> and <code>pagehide</code>
- [<code>unload</code> event](https://developer.mozilla.org/en-US/docs/Web/API/Window/unload_event) I was going to suggest the <code>unload</code> event but it is deprecated alternatives are mention in the Usage Notes section.
- [<code>beforeunload</code> event](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event) This mentions when the dialog box appears.

**Comment by bilogic** [comment]
@SakkyOP Hmm, what will happen if F5 is pressed in your approach?

**Comment by damencho** [comment]
You cannot execute async code in those events, it is unreliable.

**Comment by bilogic** [comment]
@damencho then are we better off with blocking F5 and ctrl+R?

**Comment by damencho** [comment]
I don't think we will want to block that.

**Comment by damencho** [comment]
You can configure it on your deployment via some of the empty file hooks in index.html.

**Comment by bilogic** [comment]
@damencho any examples of file hooks in index.html?

**Comment by damencho** [comment]
https://github.com/jitsi/jitsi-meet/blob/master/head.html
https://github.com/jitsi/jitsi-meet/blob/master/plugin.head.html

The content of these files are included in index.html
https://github.com/jitsi/jitsi-meet/blob/06365271463133ad304ceb9f94323952b6c4715d/index.html#L3

You can use an nginx rule to override the file location, so you can put the files in /etc/ folder or somewhere else which is different than /use/share/jitsi-meet so upgrades will not overwrite your changes.
Similar to this https://github.com/jitsi/jitsi-meet/blob/06365271463133ad304ceb9f94323952b6c4715d/doc/debian/jitsi-meet/jitsi-meet.example#L84

**Comment by bilogic** [comment]
Thanks, I'm using https://github.com/jitsi/docker-jitsi-meet, trying to map the files you mentioned, but the default does not expose any filesystem to the host

**Comment by damencho** [comment]
Yeah we need to add custom meet.conf where you can add the aliases to the files in /config and put the files in config. Any PRs are welcome.

**Comment by bilogic** [comment]
I will probably modify `docker-jitsi-meet`, I found and added a minimal script to `head.html` that was able to block F5 and ctrl+r
```
<script>
document.addEventListener('keydown', (e) => {
    e = e || window.event;
    if(e.keyCode == 116  || (e.ctrlKey && e.keyCode == 82) ){
        e.preventDefault();
    }
});
</script>
```
